File: tmp/sys/SwordOffer/src/main/scala/com/scala/offer/Problem28/Nqueens.scala
package com.scala.offer.Problem28

/**
 * Created by Lemonjing on 2018/6/2.
 * Github: Lemonjing
 *
 * Problem28 相关 N皇后
 */
object Nqueens {

  var sum = 0
  var _n = 0
  var cols: Array[Int] = _

  def queens(n: Int): Unit = {
    _n = n
    cols = new Array[Int](_n)
    backtrack(0)
  }

  private def backtrack(k: Int): Unit = {
    if (k == _n) {
      sum += 1
      println(s"第${sum}种方法: ")
      for (i <- 0 until _n) {
        println(s"[$i,${cols(i)}]")
      }
    } else {
      for (i <- 0 until _n) {
        cols(k) = i
        if (place(k)) {
          backtrack(k + 1)
        }
      }
    }
  }

  private def place(k: Int): Boolean = {
    for (j <- 0 until k) {
      if (cols(j) == cols(k) || math.abs(cols(j) - cols(k)) == math.abs(j - k)) {
        return false
      }
    }
    true
  }
}

File: tmp/sys/SwordOffer/src/main/java/com/rann/offer/problem28/NQueens.java
package com.rann.offer.problem28;

public class NQueens {

    private int[] cols;
    private int sum;

    public void queens(int n) {
        cols = new int[n];
        recursiveBacktrack(0, n);
        System.out.println("放置方式共：" + sum + "种");
    }

    private void recursiveBacktrack(int k, int n) {
        if (k == n) {
            sum++;
            System.out.println("第" + sum + "种方法");
            for (int i = 0; i < n; i++) {
                System.out.print(i + "行" + cols[i] + "列" + ",");
            }
            System.out.println();

        } else {
            for (int i = 0; i < n; i++) {
                cols[k] = i;
                if (place(k)) {
                    recursiveBacktrack(k + 1, n);
                }
            }
        }
    }

    private boolean place(int i) {
        for (int j = 0; j < i; j++) {
            if (cols[i] == cols[j] || Math.abs(cols[i] - cols[j]) == Math.abs(i - j)) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        new NQueens().queens(8);
    }
}
